In [1]:
import pandas as pd

In [2]:
products = pd.read_csv("../data/raw/product_info.csv")

In [3]:
products.head()

,product_id,product_name,brand_id,brand_name,loves_count,rating,reviews,size,variation_type,variation_value,...,online_only,out_of_stock,sephora_exclusive,highlights,primary_category,secondary_category,tertiary_category,child_count,child_max_price,child_min_price
0,P473671,Fragrance Discovery Set,6342,19-69,6320,3.6364,11.0,NaN,NaN,NaN,...,1,0,0,"['Unisex/ Genderless Scent', 'Warm &Spicy Scen...",Fragrance,Value & Gift Sets,Perfume Gift Sets,0,NaN,NaN
1,P473668,La Habana Eau de Parfum,6342,19-69,3827,4.1538,13.0,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,85.0,30.0
2,P473662,Rainbow Bar Eau de Parfum,6342,19-69,3253,4.2500,16.0,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0
3,P473660,Kasbah Eau de Parfum,6342,19-69,3018,4.4762,21.0,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0
4,P473658,Purple Haze Eau de Parfum,6342,19-69,2691,3.2308,13.0,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0


In [4]:
products.shape

(8494, 27)

In [5]:
products.columns.tolist()

['product_id',
 'product_name',
 'brand_id',
 'brand_name',
 'loves_count',
 'rating',
 'reviews',
 'size',
 'variation_type',
 'variation_value',
 'variation_desc',
 'ingredients',
 'price_usd',
 'value_price_usd',
 'sale_price_usd',
 'limited_edition',
 'new',
 'online_only',
 'out_of_stock',
 'sephora_exclusive',
 'highlights',
 'primary_category',
 'secondary_category',
 'tertiary_category',
 'child_count',
 'child_max_price',
 'child_min_price']

In [6]:
products.info()

<class 'pandas.DataFrame'>
RangeIndex: 8494 entries, 0 to 8493
Data columns (total 27 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   product_id          8494 non-null   str    
 1   product_name        8494 non-null   str    
 2   brand_id            8494 non-null   int64  
 3   brand_name          8494 non-null   str    
 4   loves_count         8494 non-null   int64  
 5   rating              8216 non-null   float64
 6   reviews             8216 non-null   float64
 7   size                6863 non-null   str    
 8   variation_type      7050 non-null   str    
 9   variation_value     6896 non-null   str    
 10  variation_desc      1250 non-null   str    
 11  ingredients         7549 non-null   str    
 12  price_usd           8494 non-null   float64
 13  value_price_usd     451 non-null    float64
 14  sale_price_usd      270 non-null    float64
 15  limited_edition     8494 non-null   int64  
 16  new              

In [7]:
products.isnull().sum()

product_id               0
product_name             0
brand_id                 0
brand_name               0
loves_count              0
rating                 278
reviews                278
size                  1631
variation_type        1444
variation_value       1598
variation_desc        7244
ingredients            945
price_usd                0
value_price_usd       8043
sale_price_usd        8224
limited_edition          0
new                      0
online_only              0
out_of_stock             0
sephora_exclusive        0
highlights            2207
primary_category         0
secondary_category       8
tertiary_category      990
child_count              0
child_max_price       5740
child_min_price       5740
dtype: int64

In [8]:
products["primary_category"].value_counts()

primary_category
Skincare           2420
Makeup             2369
Hair               1464
Fragrance          1432
Bath & Body         405
Mini Size           288
Men                  60
Tools & Brushes      52
Gifts                 4
Name: count, dtype: int64

In [9]:
products["brand_name"].nunique()

304

In [10]:
products["primary_category"].nunique()

9

doing all this bc need to know what the dataset actually contains and how 
trustworthy/useful the dataset is before we start making charts or recommendations

# important info: 

8494 products
27 cols
304 brands
9 primary categories
    - skincare: 2420
    - makeup: 2369
    - hair: 1464
    - fragrance: 1432
    - bath & body: 405
    - mini size: 288
    - men: 60
    - tools and brushes: 52
    - gifts: 4

this is accessing the products dataset... next have to look at the reviews sample dataset 
mainly check if theres also a product id to join the sets on

In [12]:
reviews_sample = pd.read_csv("../data/raw/reviews_0-250.csv")

/var/folders/gw/b8128pz11k31ky_873j0sxsc0000gn/T/ipykernel_36629/2855595699.py:1: DtypeWarning: Columns (0: author_id) have mixed types. Specify dtype option on import or set low_memory=False.
  reviews_sample = pd.read_csv("../data/raw/reviews_0-250.csv")


In [13]:
reviews_sample.head()

,Unnamed: 0,author_id,rating,is_recommended,helpfulness,total_feedback_count,total_neg_feedback_count,total_pos_feedback_count,submission_time,review_text,review_title,skin_tone,eye_color,skin_type,hair_color,product_id,product_name,brand_name,price_usd
0,0,1741593524,5,1.0,1.0,2,0,2,2023-02-01,I use this with the Nudestix “Citrus Clean Bal...,Taught me how to double cleanse!,NaN,brown,dry,black,P504322,Gentle Hydra-Gel Face Cleanser,NUDESTIX,19.0
1,1,31423088263,1,0.0,NaN,0,0,0,2023-03-21,I bought this lip mask after reading the revie...,Disappointed,NaN,NaN,NaN,NaN,P420652,Lip Sleeping Mask Intense Hydration with Vitam...,LANEIGE,24.0
2,2,5061282401,5,1.0,NaN,0,0,0,2023-03-21,My review title says it all! I get so excited ...,New Favorite Routine,light,brown,dry,blonde,P420652,Lip Sleeping Mask Intense Hydration with Vitam...,LANEIGE,24.0
3,3,6083038851,5,1.0,NaN,0,0,0,2023-03-20,I’ve always loved this formula for a long time...,Can't go wrong with any of them,NaN,brown,combination,black,P420652,Lip Sleeping Mask Intense Hydration with Vitam...,LANEIGE,24.0
4,4,47056667835,5,1.0,NaN,0,0,0,2023-03-20,"If you have dry cracked lips, this is a must h...",A must have !!!,light,hazel,combination,NaN,P420652,Lip Sleeping Mask Intense Hydration with Vitam...,LANEIGE,24.0


In [14]:
reviews_sample.shape

(602130, 19)

In [15]:
reviews_sample.columns.tolist()

['Unnamed: 0',
 'author_id',
 'rating',
 'is_recommended',
 'helpfulness',
 'total_feedback_count',
 'total_neg_feedback_count',
 'total_pos_feedback_count',
 'submission_time',
 'review_text',
 'review_title',
 'skin_tone',
 'eye_color',
 'skin_type',
 'hair_color',
 'product_id',
 'product_name',
 'brand_name',
 'price_usd']

In [16]:
reviews_sample.info()

<class 'pandas.DataFrame'>
RangeIndex: 602130 entries, 0 to 602129
Data columns (total 19 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Unnamed: 0                602130 non-null  int64  
 1   author_id                 602130 non-null  object 
 2   rating                    602130 non-null  int64  
 3   is_recommended            484644 non-null  float64
 4   helpfulness               270298 non-null  float64
 5   total_feedback_count      602130 non-null  int64  
 6   total_neg_feedback_count  602130 non-null  int64  
 7   total_pos_feedback_count  602130 non-null  int64  
 8   submission_time           602130 non-null  str    
 9   review_text               601131 non-null  str    
 10  review_title              435119 non-null  str    
 11  skin_tone                 496074 non-null  str    
 12  eye_color                 463642 non-null  str    
 13  skin_type                 527447 non-null  str    
 14 

In [17]:
reviews_sample.isnull().sum()

Unnamed: 0                       0
author_id                        0
rating                           0
is_recommended              117486
helpfulness                 331832
total_feedback_count             0
total_neg_feedback_count         0
total_pos_feedback_count         0
submission_time                  0
review_text                    999
review_title                167011
skin_tone                   106056
eye_color                   138488
skin_type                    74683
hair_color                  141081
product_id                       0
product_name                     0
brand_name                       0
price_usd                        0
dtype: int64

In [18]:
reviews_sample["product_id"].nunique()

250

In [19]:
reviews_sample["brand_name"].nunique()

60

In [20]:
reviews_sample["product_name"].nunique()

249

In [22]:
reviews_sample["rating"].value_counts().sort_index()

rating
1     32903
2     29253
3     44018
4    106956
5    389000
Name: count, dtype: int64

In [23]:
reviews_sample["submission_time"].min(), reviews_sample["submission_time"].max()

('2008-08-28', '2023-03-21')

# initial Review Data Findings

- `reviews_0-250.csv` contains 602,130 reviews.
- represents 250 unique product IDs across 60 brands.
- review dates range from 2008 to 2023.
- ratings range from 1 to 5, with 5-star reviews being the most common.
- `product_id` is available, allowing review data to be connected to `product_info.csv`.
- review text is almost completely populated.

In [24]:
review_files = [
    "../data/raw/reviews_0-250.csv",
    "../data/raw/reviews_250-500.csv",
    "../data/raw/reviews_500-750.csv",
    "../data/raw/reviews_750-1250.csv",
    "../data/raw/reviews_1250-end.csv"
]

In [25]:
for file in review_files:
    df = pd.read_csv(file, usecols=["product_id", "brand_name"])
    
    print(file)
    print("Rows:", len(df))
    print("Unique products:", df["product_id"].nunique())
    print("Unique brands:", df["brand_name"].nunique())
    print()

../data/raw/reviews_0-250.csv
Rows: 602130
Unique products: 250
Unique brands: 60

../data/raw/reviews_250-500.csv
Rows: 206725
Unique products: 249
Unique brands: 65

../data/raw/reviews_500-750.csv
Rows: 116262
Unique products: 249
Unique brands: 79

../data/raw/reviews_750-1250.csv
Rows: 119317
Unique products: 499
Unique brands: 107

../data/raw/reviews_1250-end.csv
Rows: 49977
Unique products: 1104
Unique brands: 122



In [26]:
review_dfs = []

for file in review_files:
    temp = pd.read_csv(file)
    review_dfs.append(temp)

reviews = pd.concat(review_dfs, ignore_index=True)

/var/folders/gw/b8128pz11k31ky_873j0sxsc0000gn/T/ipykernel_36629/2397321834.py:4: DtypeWarning: Columns (0: author_id) have mixed types. Specify dtype option on import or set low_memory=False.
  temp = pd.read_csv(file)


In [27]:
reviews.shape

(1094411, 19)

In [28]:
reviews["product_id"].nunique()

2351

In [29]:
reviews["brand_name"].nunique()

142

In [30]:
reviews.duplicated().sum()

np.int64(0)

In [31]:
reviews["product_id"].isin(products["product_id"]).mean()

np.float64(1.0)

## dataset Scope Decision

the final review dataset contains:

- 1,094,411 customer reviews
- 2,351 unique reviewed products
- 142 brands
- No exact duplicate review rows
- 100% of review product IDs match the product information table

all five review files will be used because they provide broader product and brand coverage than any individual file

## core business questions

1. Which beauty categories and brands differ most in price, rating, and customer interest across the full Sephora catalog?
2. Within skincare, which brands and products show the strongest customer demand?
3. Which skincare products have high customer interest but relatively weak ratings?
4. What complaints appear most frequently in negative skincare reviews?
5. What qualities appear most frequently in positive skincare reviews?
6. Which combinations of demand, price, rating, and review themes suggest potential skincare product opportunities?

In [37]:
reviewed_product_ids = reviews["product_id"].unique()

reviewed_products = products[
    products["product_id"].isin(reviewed_product_ids)
]

reviewed_products["primary_category"].value_counts()

primary_category
Skincare    2351
Name: count, dtype: int64

# review coverage limitation

all 2351 products represented in customer review files r skincare

because of this:
- market-level product analysis can use the full sephora catalog of 8494 products across all skincare categories
- review sentiment, complain, priase, and oppurtunity analysis will focus specifically on skincare
- the project will not claim to have review-text coverage for makeup, hair, fragrance, or other categories